File: backend/src/task_center/context_engine/scope.py
"""ContextScope — identity surface every role context builder sees.

The scope carries identity (workflow / iteration / attempt / task ids). It
does **not** carry store handles; those live on :class:`ContextEngineDeps`
so builders do not reach into globals.

The role-specific factory classmethods (:meth:`for_planner`,
:meth:`for_generator`, etc.) document the required fields per role at the
call site: omitting one raises ``TypeError`` at call time, and strict
mypy will narrow the kwargs to their declared ``str`` types. The engine still
validates fields so direct ``ContextScope(...)`` construction is covered.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Self

from task_center.context_engine.exceptions import RecipeScopeError

ScopeField = Literal["workflow_id", "iteration_id", "attempt_id", "task_id"]


@dataclass(frozen=True, slots=True)
class ContextScope:
    """Identity surface threaded through resolver + engine + builders."""

    role: Literal["planner", "generator", "reducer"] = "planner"
    workflow_id: str | None = None

    # Optional identity fields — recipes declare which of these they need.
    iteration_id: str | None = None
    attempt_id: str | None = None
    task_id: str | None = None

    def require_field(self, field: ScopeField) -> str:
        """Return one required identity field as a non-optional string."""
        value = getattr(self, field)
        if not isinstance(value, str):
            raise RecipeScopeError(f"ContextScope is missing required field: {field!r}")
        return value

    # ---- Role-specific factory shortcuts -------------------------------
    #
    # Each factory takes ONLY the required fields for that launch role as
    # positional/keyword args. Missing a required field is a static error
    # instead of a runtime assert. The defaults flow through the dataclass
    # for any optional fields the role might inspect.

    @classmethod
    def for_planner(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
    ) -> Self:
        """Scope shape required by the planner context."""
        return cls(
            role="planner",
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
        )

    @classmethod
    def for_generator(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
        task_id: str,
    ) -> Self:
        """Scope shape required by the generator context."""
        return cls(
            role="generator",
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
            task_id=task_id,
        )

    @classmethod
    def for_reducer(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
        task_id: str,
    ) -> Self:
        """Scope shape required by the reducer context."""
        return cls(
            role="reducer",
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
            task_id=task_id,
        )

File: agent-core/crates/eos-workflow/src/context/scope.rs
use eos_state::{AttemptId, IterationId, TaskId, WorkflowId};

use super::ContextRole;

/// Identity a context builder reads, keyed by launch role so each role carries
/// exactly the ids it requires. Constructed only through the `for_*`
/// constructors, which makes an id/role mismatch unrepresentable.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ContextScope {
    /// Planner launch: workflow + iteration + attempt.
    Planner {
        /// Workflow id.
        workflow_id: WorkflowId,
        /// Iteration id.
        iteration_id: IterationId,
        /// Attempt id.
        attempt_id: AttemptId,
    },
    /// Generator launch: planner ids plus the assigned task.
    Generator {
        /// Workflow id.
        workflow_id: WorkflowId,
        /// Iteration id.
        iteration_id: IterationId,
        /// Attempt id.
        attempt_id: AttemptId,
        /// Assigned task id.
        task_id: TaskId,
    },
    /// Reducer launch: planner ids plus the assigned task.
    Reducer {
        /// Workflow id.
        workflow_id: WorkflowId,
        /// Iteration id.
        iteration_id: IterationId,
        /// Attempt id.
        attempt_id: AttemptId,
        /// Assigned task id.
        task_id: TaskId,
    },
}

impl ContextScope {
    /// Scope for a planner launch.
    #[must_use]
    pub fn for_planner(
        workflow_id: WorkflowId,
        iteration_id: IterationId,
        attempt_id: AttemptId,
    ) -> Self {
        Self::Planner {
            workflow_id,
            iteration_id,
            attempt_id,
        }
    }

    /// Scope for a generator launch.
    #[must_use]
    pub fn for_generator(
        workflow_id: WorkflowId,
        iteration_id: IterationId,
        attempt_id: AttemptId,
        task_id: TaskId,
    ) -> Self {
        Self::Generator {
            workflow_id,
            iteration_id,
            attempt_id,
            task_id,
        }
    }

    /// Scope for a reducer launch.
    #[must_use]
    pub fn for_reducer(
        workflow_id: WorkflowId,
        iteration_id: IterationId,
        attempt_id: AttemptId,
        task_id: TaskId,
    ) -> Self {
        Self::Reducer {
            workflow_id,
            iteration_id,
            attempt_id,
            task_id,
        }
    }

    /// The launch role this scope was built for.
    #[must_use]
    pub fn role(&self) -> ContextRole {
        match self {
            Self::Planner { .. } => ContextRole::Planner,
            Self::Generator { .. } => ContextRole::Generator,
            Self::Reducer { .. } => ContextRole::Reducer,
        }
    }
}
